File: src/com/chbachman/cminus/representation/statement/Print.java
package com.chbachman.cminus.representation.statement;

import com.chbachman.cminus.CMinusParser;
import com.chbachman.cminus.representation.Scope;
import com.chbachman.cminus.representation.Type;
import com.chbachman.cminus.representation.value.Value;

/**
 * Created by Chandler on 4/12/17.
 */
public class Print implements Statement {

    public final Value value;
    public final String modifier;

    public Print(CMinusParser.PrintContext ctx, Scope scope) {
        this.value = Value.parse(ctx.value(), scope);

        switch (Type.Native.get(value.type())) {
            case INT: modifier = "%d"; break;
            case STRING: modifier = "%s"; break;
            case BOOL: modifier = "%s"; break;
            default: throw new RuntimeException("Cannot print value " + value.value());
        }
    }

    @Override
    public String code() {
        if (value.type() == Type.Native.BOOL.type) {
            return "printf(\"" + modifier + "\\n\", " + value.value() + " ? \"true\" : \"false\");";
        }
        return "printf(\"" + modifier + "\\n\", " + value.value() + ");";
    }
}

File: src/main/java/com/chbachman/cminus/representation/statement/Print.kt
package com.chbachman.cminus.representation.statement

import com.chbachman.cminus.gen.CMinusParser
import com.chbachman.cminus.representation.Parser
import com.chbachman.cminus.representation.Scope
import com.chbachman.cminus.representation.Type
import com.chbachman.cminus.representation.Typed
import com.chbachman.cminus.representation.value.Value

/**
 * Created by Chandler on 4/12/17.
 */
class Print(ctx: CMinusParser.PrintContext, scope: Scope) : Statement {

    val value = Parser.parse(ctx.value(), scope)
    val modifier: String

    init {
        when (value.type) {
            Type.Native.INT.type -> modifier = "%d"
            Type.Native.STRING.type -> modifier = "%s"
            Type.Native.BOOL.type -> modifier = "%s"
            else -> throw RuntimeException("Cannot print value " + value.value())
        }
    }

    override fun code(): String {
        if (value.type == Type.Native.BOOL.type) {
            return "printf(\"$modifier\\n\", ${value.value()} ? \"true\" : \"false\");"
        }
        return "printf(\"$modifier\\n\", ${value.value()});"
    }
}
